File: 100-same-tree/100-same-tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int val=0, TreeNode left=null, TreeNode right=null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
public class Solution {
    
    /**
     * IDEA:
     * - Use recursion to traverse all nodes in the two trees simulataneously. The recursive
     *   function should take in the current TreeNode of each tree (each is incremented in the 
     *   previous recursive call). At the start of each recursive call, check if the two 
     *   TreeNodes that were passed in are the same. If they are, continue with the recursion;
     *   otherwise, return false. If the whole tree is searched without a premature return, 
     *   the trees are the same, return true.
     */
    public bool IsSameTree(TreeNode p, TreeNode q) {
        
        return recursiveIsSameTree(p, q);
    }
    
    private bool recursiveIsSameTree(TreeNode p, TreeNode q) {
        Console.WriteLine("AT: recusriveIsSameTree()");
        
        // BASE CASE 2: End of branch reached, return.
        if (p == null && q == null) { return true; }
        
        Console.WriteLine("Case 1.1 Passed");
        
        if (p == null || q == null) { return false; }
        
        Console.WriteLine("Case 1.2 Passed");
        Console.WriteLine($"p: {p.val}, q: {q.val}");
        
        // BASE CASE 2: Nodes are not equal; therefore, trees are not equal.
        if (p.val != q.val) { return false; }
        
        Console.WriteLine("Case 2 Passed");
        
        return recursiveIsSameTree(p.left, q.left) && recursiveIsSameTree(p.right, q.right);
    }
}

File: LeetCode/0100-same-tree/100-same-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        pStack = []
        qStack = []

        pStack.append(p)
        qStack.append(q)

        # If the trees have a different number of nodes, a comparison will always be made while looping between a Null and Non-Null pairs of nodes. Thus, we shouldn't have to worry about incorrectly returning True here. That is, if one stack runs out of nodes, a None node from the empty stack will be compared to a non-None node from the other stack before the iteration in which the stack was emptied will occur and cause us to return False.
        while pStack and qStack:
            pNode = pStack.pop()
            qNode = qStack.pop()

            if (not pNode and qNode) or (pNode and not qNode):
                return False

            if (pNode and qNode) and (pNode.val != qNode.val):
                return False

            if pNode:
                pStack.append(pNode.right)
                pStack.append(pNode.left)
            if qNode:
                qStack.append(qNode.right)
                qStack.append(qNode.left)

        return True


# ***** IDEAS *****
# IDEA 1:
# > Iteratively traverse both trees using DFS. (although, it doesn't really matter what traversal method we use)
#   > Note that it must be iterative. If we use recursion, we will have to find a way to pass nodes between the different recursive calls that are created. At least I think we would have to.
#
# ***** OUTLINE *****
# Instantiate two stacks: pStack and qStack
# Push the root node of each tree to their resepecitve stacks.
# Loop over all of the nodes in the tree
#     Pop the current node off of each stack.
#     If pCurrent != qCurrent, return False
#     Push the left and right children of both trees to the stack.
# Return True